### ISSUE TITLE
Add support for async "is" function in "when"

### ISSUE THREAD

**Issue by igo** [jquense/yup#2211]
Add support for async `is` function so that async functions can be called inside.
```ts
.when(['$prisma'], {
    is: async (prisma: PrismaClient) => {
      return await prisma.entity.count({...});
    }
})
```
Right now seems that if there is an `async` it always evaluates to true.

**Comment by DaddyWarbucks** [comment]
I don't think this will be possible in current versions of yup. The `.when()` method is called at schema build time and doesn't support async functions. But you are justing using `.when()` with a `$` where you are trying to get some context, not necessarily changing a schema shape based on another property. You likely want to use `.test()` where you can access the context and do what you want. `when()` is generally for changing a schema shape. You have access to context there, but I think you are looking for `.test()`
